examples/petsc/multigrid.c

```diff
@@ -429,24 +428,42 @@ int main(int argc, char **argv) {
     }
   }
 
-  // Setup dummy SNES for AMG coarse solve
-  ierr = SNESCreate(comm, &snes_dummy); CHKERRQ(ierr);
-  ierr = SNESSetDM(snes_dummy, dm[0]); CHKERRQ(ierr);
-  ierr = SNESSetSolution(snes_dummy, X[0]); CHKERRQ(ierr);
-
-  // -- Jacobian matrix
-  ierr = DMSetMatType(dm[0], MATAIJ); CHKERRQ(ierr);
+  // Assemble coarse grid Jacobian for AMG (or other sparse matrix) solve
   ierr = DMCreateMatrix(dm[0], &mat_coarse); CHKERRQ(ierr);
-  ierr = SNESSetJacobian(snes_dummy, mat_coarse, mat_coarse, NULL,
-                         NULL); CHKERRQ(ierr);
-
-  // -- Residual evaluation function
-  ierr = SNESSetFunction(snes_dummy, X[0], FormResidual_Ceed,
-                         user_O[0]); CHKERRQ(ierr);
 
-  // -- Form Jacobian
-  ierr = SNESComputeJacobianDefaultColor(snes_dummy, X[0], mat_O[0],
-                                         mat_coarse, NULL); CHKERRQ(ierr);
+  if (PETSC_VERSION_GE(3,16,0)) { // Assemble matrix analytically
+    CeedInt num_entries, *rows, *cols;
+    CeedVector coo_values;
+    CeedOperatorLinearAssembleSymbolic(user_O[0]->op, &num_entries, &rows, &cols);
+    ISLocalToGlobalMapping ltog_row, ltog_col;
+    ierr = MatGetLocalToGlobalMapping(mat_coarse, &ltog_row, &ltog_col); CHKERRQ(ierr);
+    ierr = ISLocalToGlobalMappingApply(ltog_row, num_entries, rows, rows); CHKERRQ(ierr);
+    ierr = ISLocalToGlobalMappingApply(ltog_col, num_entries, cols, cols); CHKERRQ(ierr);
+    ierr = MatSetPreallocationCOO(mat_coarse, num_entries, rows, cols); CHKERRQ(ierr);
+    free(rows);
+    free(cols);
+    CeedVectorCreate(ceed, num_entries, &coo_values);
+    CeedOperatorLinearAssemble(user_O[0]->op, coo_values);
+    const CeedScalar *values;
+    CeedVectorGetArrayRead(coo_values, CEED_MEM_HOST, &values);
+    ierr = MatSetValuesCOO(mat_coarse, values, ADD_VALUES); CHKERRQ(ierr);
+    CeedVectorRestoreArrayRead(coo_values, &values);
+    CeedVectorDestroy(&coo_values);
```

Comment:
@jeremylt I'd like to apply this pattern in Solids. It currently requires [my branch in PETSc](https://gitlab.com/petsc/petsc/-/merge_requests/4293#note_674007495) to correctly handle negative indices. This should save us significant setup time.